### PR TITLE
[Feature] Separate losses

### PR DESCRIPTION
**Pull request by MateuszGuzek**
## Description

TD3 should have the option of tying or not the parameters of the actor and qvalue network as in https://github.com/pytorch/rl/pull/1037. We should test whether the loss is backprop on qvalue net or not in the tests

- [x] TD3 
- [x] Ditto for DDPGLoss
- [x] Ditto for IQLLoss
- [x] Ditto for reinforce
- [x] Ditto for A2C
- [x] Ditto for SAC
- [x] Ditto for REDQ (+ deprecated version)

## Motivation and Context

The first 7 tasks of https://github.com/pytorch/rl/issues/1058

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (update in the documentation)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: test/test_cost.py [hunk 1]
```diff
@@ -1106,6 +1108,7 @@ def _create_seq_mock_data_td3(
     @pytest.mark.parametrize("noise_clip", [0.1, 1.0])
     @pytest.mark.parametrize("td_est", list(ValueEstimators) + [None])
     @pytest.mark.parametrize("use_action_spec", [True, False])
+    @pytest.mark.parametrize("separate_losses", [False, True])
```
Comment (vmoens) [inline test/test_cost.py:1111]: since we test it just after I would remove it from here

File: test/test_cost.py [hunk 2]
```diff
@@ -771,7 +771,8 @@ def _create_seq_mock_data_ddpg(
     @pytest.mark.parametrize("device", get_available_devices())
     @pytest.mark.parametrize("delay_actor,delay_value", [(False, False), (True, True)])
     @pytest.mark.parametrize("td_est", list(ValueEstimators) + [None])
-    def test_ddpg(self, delay_actor, delay_value, device, td_est):
+    @pytest.mark.parametrize("separate_losses", [False, True])
```
Comment (vmoens) [inline test/test_cost.py:774]: This should be tested separately

File: test/test_cost.py [hunk 3]
```diff
@@ -1191,6 +1196,100 @@ def test_td3(
...
+    @pytest.mark.skipif(not _has_functorch, reason="functorch not installed")
+    @pytest.mark.parametrize("device", get_available_devices())
+    @pytest.mark.parametrize(
+        "delay_actor, delay_qvalue", [(False, False), (True, True)]
+    )
+    @pytest.mark.parametrize("td_est", list(ValueEstimators) + [None])
+    @pytest.mark.parametrize("policy_noise", [0.1, 1.0])
+    @pytest.mark.parametrize("noise_clip", [0.1, 1.0])
+    @pytest.mark.parametrize("use_action_spec", [True, False])
+    @pytest.mark.parametrize("separate_losses", [False, True])
```
Comment (vmoens) [inline test/test_cost.py:1208]: let's just keep separate loss and use defaults for all others

File: test/test_cost.py [hunk 4]
```diff
@@ -1460,6 +1559,7 @@ def _create_seq_mock_data_sac(
     @pytest.mark.parametrize("num_qvalue", [1, 2, 4, 8])
     @pytest.mark.parametrize("device", get_available_devices())
     @pytest.mark.parametrize("td_est", list(ValueEstimators) + [None])
+    @pytest.mark.parametrize("separate_losses", [False, True])
```
Comment (vmoens) [inline test/test_cost.py:1562]: since we test it just after I would remove it from here

File: test/test_cost.py [hunk 5]
```diff
@@ -2843,7 +2945,8 @@ def _create_seq_mock_data_ppo(
     @pytest.mark.parametrize("advantage", ("gae", "td", "td_lambda", None))
     @pytest.mark.parametrize("device", get_available_devices())
     @pytest.mark.parametrize("td_est", list(ValueEstimators) + [None])
-    def test_ppo(self, loss_class, device, gradient_mode, advantage, td_est):
+    @pytest.mark.parametrize("separate_losses", [False, True])
```
Comment (vmoens) [inline test/test_cost.py:2948]: since we test it just after I would remove it from here

File: test/test_cost.py [hunk 6]
```diff
@@ -1264,6 +1324,76 @@ def test_td3(
...
+        # td = self._create_mock_data_td3(device=device)
+        actor, value, common, td = self._create_mock_common_layer_setup(n_act=n_act)
+        loss_fn = TD3Loss(
+            actor,
+            value,
+            action_spec=BoundedTensorSpec(shape=(n_act,), minimum=-1, maximum=1),
+            loss_function="l2",
+            separate_losses=separate_losses,
+        )
+        with _check_td_steady(td):
```
Comment (vmoens) [inline test/test_cost.py:1348]: we don't need this check here

File: test/test_cost.py [hunk 7]
```diff
@@ -1264,6 +1324,76 @@ def test_td3(
...
+    def test_td3_separate_losses(
+        self,
+        device,
+        separate_losses,
+        n_act=4,
+    ):
+        torch.manual_seed(self.seed)
+        # actor = self._create_mock_actor(device=device)
+        # value = self._create_mock_value(device=device)
+        # td = self._create_mock_data_td3(device=device)
```
Comment (vmoens) [inline test/test_cost.py:1339]: to be removed

File: test/test_cost.py [hunk 8]
```diff
@@ -2724,18 +3368,36 @@ def test_redq(self, delay_qvalue, num_qvalue, device, td_est):
                     )
                 )
             elif k == "loss_qvalue":
-                assert all(
+                assert not any(
```
Comment (vmoens) [inline test/test_cost.py:3371]: has the default behaviour of REDQ changed?

**Review by vmoens — APPROVED** [review]
LGTM fantastic work
Can you just comment on REDQ: has the default behaviour changed?

File: torchrl/objectives/redq.py [hunk 9]
```diff
@@ -253,6 +257,7 @@ def __init__(
         gSDE: bool = False,
         gamma: float = None,
         priority_key: str = None,
+        separate_losses: bool = True,
```
Comment (MateuszGuzek) [inline torchrl/objectives/redq.py:260]: I had to set redq defaults to true, otherwise the other existing tests failed.

Comment (MateuszGuzek) [reply torchrl/objectives/redq.py:260]: If I remove the REDQ tests and make them more lightweight, like the other tests I did (only focused on checking the gradients), I can change the defaults to False here and all tests pass. I will make a commit with this change.

File: test/test_cost.py (hunk 8 above)
Comment (MateuszGuzek) [reply test/test_cost.py:3371]: I restored the defaults, I changed it to False
